### DISCUSSION TITLE
IDL dynamic libraries with GDL

### DISCUSSION THREAD

**Discussion by nabajour** [gnudatalanguage/gdl#1192] [Q&A]
Hi,
I got an IDL code that I should port to another language. I'm all new to IDL, currently, I'm trying to run that script in GDL to look at the output and compare the code's behaviour. 

The IDL code uses NAIF SPICE toolkit [1], with it's IDL interface, "icy". From what I understand, the toolkit uses DLM, and the code accesses it with the DLM_register function. And to compile the IDL interface, it needs an "export.h" header that's provided by IDL that don't exist in GDL (and that seems to declare structures that don't exist in GDL).

So my question is this: is using DLM stuff from IDL supported by GDL ? Any idea if someone uses SPICE from GDL? 
If not, as the code seems to only use a handful of functions, it might be possible to just wrap the ones I need and call them from the spice C interface with CALL_EXTERNAL, but I don't know how much work the IDL code does and what other incompatibilites lurk in the code.

[1] https://naif.jpl.nasa.gov/naif/toolkit.html

**Comment by alaingdl** [comment]
Not sure to exactly understand what you are asking for, because of my limited technical background

Do you mean : can you linked directly a .o or .so created for IDL in GDL ? Yes, I did that for quiet large and complex libs. in C without trouble. (and we read the raw Planck HFI data in GDL :smile: without trouble). You can have a look to `test_call_external.pro` and `libtest_ce.cpp` in the testsuite/ The cpp file is maybe provideing exactly what you are looking for.

I started to play with ICY/SPICE/NAIF, and yes I succeed to compile the C code ... but I did not located a tuto or demo up to now to test the IDL side.

**Reply by nabajour** [reply]
Hi, 

Thanks for the answer. I'm still learning and reading stuff about ICY/SPICE and IDL DLMs, sorry if my question is a bit fuzzy, it reflects my confusion :-P  I'm trying to find the best way to call those functions from CPICE wrapped by ICY without reinventing the wheel.

So from what I understood, CSPICE is the C library (The C code in the SPICE IDL distribution is exactly the same as the C module). Then, there is ICY, which is the C part of the library and some *.pro files, providing facilities for IDL to find functions, and call them. In the icy.c file of the SPICE/IDL it seems to have code to register the functions from ICY to IDL. This file seems to also contains a lot of wrapper code for all the CSPICE functions that checks the arguments and allocates memory for the returned data from CSPICE functions. 
In IDL, this seems to work that the user can call the functions in the same way as IDL functions, using DLM_register  (This type of magic is described here: https://www.l3harrisgeospatial.com/docs/dlm.html ).  This doesn't seem to exist in GDL. 

So from my understanding of GDL and your answer, I have to load the functions myself, with CALL_EXTERNAL.
And I guess one option would be to use the CSPICE library and directly call the C code functions (and handle the memory allocation) from GDL? (Thus reinventing the code that's in ICY).
Would it be possible to call the spice wrapper functions from icy.so with CALL_EXTERNAL and thus gain the advantage of having all the allocation done by those wrapper functions? They might depend on some facilities of IDL that are not provided by GDL.

It looks like I can have access to a machine with IDL at my Uni, so this might solve part of my problem, but it would make things easier if I can run it on my local machine with GDL. 

I'll try to come up with a minimum working example to call some ICY/SPICE functions.

Thanks

**Comment by alaingdl** [comment]
going ahead ...
I made 2 small changes in `icy/src/icy/mkprodct.csh`
```
   set INCLUDESOPTIONS  = "-I/usr/local/harris/idl/external -I../../include"
   set WARNINGS1        = "-Wall -Wundef"
```
and succeed to compile ...

**Reply by nabajour** [reply]
I don't have IDL on my machine, so the compilation of icy.so fails because it doesn't find the "export.h" function. (I think this one defines IDL_Load() functions and probably some structures to pass functions prototypes and definitions between the lib and IDL). 
There is a precompiled icy.so file in the SPICE distribution, so maybe I can use this, but if it needs some symbols from IDL that are not in GDL, I think I'm stuck and need to call CSPICE directly.

**Comment by HuwMorgan** [comment]
Was this issue resolved? I've been seeking a similar solution, to use GDL with the SPICE routines. In GDL, I can use linkimage procedure to link to a function:
GDL> linkimage,'CSPICE_B1950','icy.so',1

But when I use call_external, GDL crashes (exits) with the following message:
GDL> a=call_external('icy.so','CSPICE_B1950')
dyld[1067]: missing symbol called
Magick: abort due to signal 6 (SIGABRT) "Abort"...
Abort trap: 6

I've searched for the error message above online, and have found several discussions relating to ImageMagick, but I'm not so sure this is a solution for our problem! Any help appreciated.

**Reply by GillesDuvert** [reply]
My experience is only that [this software](https://github.com/JMMC-OpenDev/wisard-ci) uses call_external to call a library called [optimpack](https://github.com/JMMC-OpenDev/wisard-ci)  without problems. Same with the NIKA2 instrument package (not public).
